### PR TITLE
gdb special command to print tensors

### PR DESCRIPTION
**Pull request by antocuni**
This is something which I wrote because it was useful during my debugging sessions, but I think it might be generally useful to other people as well so I took the liberty of proposing an official `pytorch-gdb` extension.

`pytorch-gdb` is a gdb script written in python. Currently, it contains only one command: `torch-tensor-repr`, which prints a human-readable repr of an `at::Tensor` object. Example:
```
Breakpoint 1, at::native::neg (self=...) at [...]/pytorch/aten/src/ATen/native/UnaryOps.cpp:520
520     Tensor neg(const Tensor& self) { return unary_op_impl(self, at::neg_out); }
(gdb) # the default repr of 'self' is not very useful
(gdb) p self
$1 = (const at::Tensor &) @0x7ffff72ed780: {impl_ = {target_ = 0x5555559df6e0}}
(gdb) torch-tensor-repr self
Python-level repr of self:
tensor([1., 2., 3., 4.], dtype=torch.float64)
```

The idea is that by having an official place where to put these things, `pytorch-gdb` will slowly grow other useful features and make the pytorch debugging experience nicer and faster.

### CODE REVIEW

File: torch/csrc/utils.cpp [hunk 1]
```diff
@@ -252,3 +252,59 @@ void THPPointer<THPStorage>::free() {
...
+    fprintf(stderr, "cannot allocate memory for the result\n");
+    goto error;
+  }
+  strcpy(result, buf);
+  Py_XDECREF(pytensor);
+  Py_XDECREF(repr);
+  PyGILState_Release(gil);
+  return result;
+
+error:
```
Comment (ezyang) [inline torch/csrc/utils.cpp:299]: Not using RAII here is kind of weird. Is there a particular reason, e.g., gdb doesn't like it when you have stack unwinding?

**Review by ezyang — APPROVED** [review]

File: torch/csrc/utils.cpp (hunk 1 above)
Comment (antocuni) [reply torch/csrc/utils.cpp:299]: the most honest answer is that I am used to write Python API code in C and I didn't think that this was C++ :man_facepalming: 

But also, I am not aware of any standard smart-pointer-like class to manage `PyObject*` in C++ with RAII. Do we have any helper for that?

Or you were suggesting do write a custom class just to handle this specific function?

**Comment by wwwwcu** [comment]
I saw the tool (pytorhc-gdb), so I want to have a try.
Following the guide, I can't get the value of at::tensor since my code walk into libtorch_cpu.so instead of UnaryOps.cpp:520
`
gdb python
GNU gdb (Ubuntu 8.1-0ubuntu3.2) 8.1.0.20180409-git
[...]
(gdb) break at::native::negative
Function "at::native::negative" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (at::native::negative) pending.
(gdb) run
Starting program: /root/anaconda3/envs/pytorch/bin/python 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Python 3.6.10 |Anaconda, Inc.| (default, May  8 2020, 02:54:21) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> t = torch.tensor([1, 2, 3, 4], dtype=torch.float64)
>>> t.negative()

**Breakpoint 1, 0x00007fffe32c5f20 in at::native::negative(at::Tensor const&)@plt () from /root/anaconda3/envs/pytorch/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so**
`
I built pytorch from source with "DEBUG=1", I'm not sure if I use the tool correctly.

**Comment by antocuni** [comment]
> I saw the tool (pytorhc-gdb), so I want to have a try.
> Following the guide, I can't get the value of at::tensor since my code walk into libtorch_cpu.so instead of UnaryOps.cpp:520

you are correct, the example was out of date. I fixed it in #61175

**Pull request closed without merging**